doc/designs/ech-api.md

```diff
@@ -0,0 +1,590 @@
+``SSL_CTX_ech_raw_decrypt()`` (for the initial and second ClientHello
+messages). When done, the caller must free the ``hrrtok`` using
+``OPENSSL_free()``.  If the caller doesn't need to support HRR, then it can
+supply NULL values for these parameters. The value of the token is the client's
+ephemeral public value, which is not sensitive having being sent in clear in
+the first ClientHello.  This value is missing from the second ClientHello but
+is needed for ECH decryption.
+
+Note that ``SSL_CTX_ech_raw_decrypt()`` only takes a ClientHello as input. If
+the flight containing the ClientHello contains other messages (e.g. a
+ChangeCipherSuite or Early data), then the caller is responsible for
+disentangling those, and for assembling a new flight containing the inner
+ClientHello.
+
+### ECH-specific Padding of server messages
+
+If a web server were to host a set of web sites, one of which had a much longer
+name than the others, the size of some TLS handshake server messages could
+expose which web site was being accessed. Similarly, if the TLS server
+certificate for one web site were significantly larger or smaller than others,
+message sizes could reveal which web site was being visited.  For these
+reasons, we provide a way to enable additional ECH-specific padding of the
+Certificate, CertificateVerify and EncryptedExtensions messages sent from the
+server to the client during the handshake.
+
+To enable ECH-specific padding, one makes a call to:
+
+```c
+    SSL_CTX_set_options(ctx, SSL_OP_ECH_SPECIFIC_PADDING);
+```
+
+The default padding scheme is to ensure the following sizes for the plaintext
+form of these messages:
+
+| Message             | Minimum Size | Size is multiple of |
+| ------------------- | ------------ | ------------------- |
+| Certificate         | 1792         | 128                 |
+| CertificateVerify   | 480          | 16                  |
+| EncryptedExtensions | 32           | 16                  |
+
+The ciphertext form of these messages, as seen on the network in the record
+layer protocol, will usually be 16 octets more, due to the AEAD tag that is
+added as part of encryption.
+
+If a server wishes to have finer-grained control of these sizes, then it can
+make use of the ``SSL_CTX_ech_set_pad_sizes()`` or ``SSL_ech_set_pad_sizes()``
+APIs. Both involve populating an ``OSSL_ECH_PAD_SIZES`` data structure as
+described below in the obvious manner.
+
+```c
+/*
+ * Fine-grained ECH-spacific padding controls for a server
+ */
+typedef struct ossl_ech_pad_sizes_st {
+    size_t cert_min; /* minimum size */
+    size_t cert_unit; /* size will be multiple of */
+    size_t certver_min; /* minimum size */
+    size_t certver_unit; /* size will be multiple of */
+    size_t ee_min; /* minimum size */
+    size_t ee_unit; /* size will be multiple of */
+} OSSL_ECH_PAD_SIZES;
+
+int SSL_CTX_ech_set_pad_sizes(SSL_CTX *ctx, OSSL_ECH_PAD_SIZES *sizes);
+int SSL_ech_set_pad_sizes(SSL *s, OSSL_ECH_PAD_SIZES *sizes);
```

Comment:
At the protocol level, there is a need for padding for a new reason (e.g. certificate size reducing the size of the anonymity set). Until recently I implemented that via callbacks in `s_server` but Viktor convinced me that it ought be done inside the library instead because otherwise the last callback added would "win" and we might not pad effectively (for ECH purposes). The APIs to set the sizes are intended so that an application could try to mimic the padding used by some other popular application(s). (I agree that that last is a bit speculative but could help in some censorship circumvention situations.)

I guess setting a large-ish value via ``SSL_set_block_padding()`` could achieve the required effect but maybe at the expense of being less efficient after the handshake.

So, I think I'd (not that strongly, but a bit:-) argue to keep the proposed ECH-specific padding, or perhaps to add a new option to only apply block padding during the handshake, or even only to the relevant handshake messages (i.e. Certificate, CertificateVerify and EncryptedExtensions).

Reply:
How about `SSL_CTX_set_block_padding_ex`/`SSL_set_block_padding_ex` functions that look like the existing non-ex variants but with an additional `uint64_t flags` argument which provides finer degree control over how they operate. You could introduce one or more flags to specify which bits of the handshake it applies to, and whether the padding occurs after the handshake is complete or not.

The different sizes for different handshake messages really seems highly speculative and it seems unnecessary in my eyes. If someone *really* wants that degree of control they could probably do it via a custom callback approach.

Reply:
Actually, that's nice, yes - better as a separate PR for master or kept in the feature branch? It probably is mostly useful for ECH but I guess could have some broader applicability and it doesn't really have to refer to any other ECH stuff.

Question though: should that allow different block_padding sizes per flag or would just one for all be enough? IOW, would we want to set say rounding-to-32 for certificate messages but rounding-to-16 for all others?

Reply:
> Actually, that's nice, yes - better as a separate PR for master or kept in the feature branch? It probably is mostly useful for ECH but I guess could have some broader applicability and it doesn't really have to refer to any other ECH stuff.

Yeah, then it makes sense to do it as a separate PR against master branch.

Reply:
> Question though: should that allow different block_padding sizes per flag or would just one for all be enough? IOW, would we want to set say rounding-to-32 for certificate messages but rounding-to-16 for all others?

I would keep it simple and just have the same rounding for all. We could always add another flag later if we want special handling for certificate messages.

Reply:
OK, I'll submit a separate PR for master along those lines in the next few days.